src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -712,23 +711,33 @@ ArrayJoinActionPtr SelectQueryExpressionAnalyzer::appendArrayJoin(ExpressionActi
     return array_join;
 }
 
-bool SelectQueryExpressionAnalyzer::appendJoinLeftKeys(ExpressionActionsChain & chain, bool only_types)
+bool SelectQueryExpressionAnalyzer::appendJoinLeftKeys(ExpressionActionsChain & chain, bool only_types, Block & block)
 {
     ExpressionActionsChain::Step & step = chain.lastStep(columns_after_array_join);
 
     getRootActions(analyzedJoin().leftKeysList(), only_types, step.actions());
+    ExpressionActionsPtr actions = std::make_shared<ExpressionActions>(step.actions());
+    actions->execute(block);
     return true;
 }
 
-JoinPtr SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain)
+JoinPtr
+SelectQueryExpressionAnalyzer::appendJoin(ExpressionActionsChain & chain, const Block & sample_block, ActionsDAGPtr & before_join_dag)
 {
-    JoinPtr table_join = makeTableJoin(*syntax->ast_join);
+    JoinCommon::JoinConvertActions converting_actions;
+    JoinPtr table_join = makeTableJoin(*syntax->ast_join, sample_block, converting_actions);
```

Comment:
I think, instead of passing `sample_block` here, we can use `chain.getLastStep().getResultColumns()`. Cause we have added actions to chain in `appendJoinLeftKeys`

Reply:
Yes, it works https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-885608312a315dd9c5ce4577294b073f1be45b4fb55eba485304861cf8e3e1edR743